### PR TITLE
Upgrade CI test frameworks and fix linking NCCL 2.12+

### PR DESCRIPTION
**Pull request by EnricoMi**
Another round of upgrading tested frameworks:
- fix MXNet versions
- upgrade torch to 1.13.1
- upgrade pyspark to 3.3.2, add 3.4.0, remove 3.2.3
- add Tensorflow 2.12.0, remove 2.9.3, upgrade to 2.11.1

### CODE REVIEW

**Comment by maxhgerlach** [comment]
@EnricoMi, would you have some bandwidth to update this to TensorFlow 2.12 (and probably PyTorch 2.0.0)? `tf-nightly` is a moving target at the moment, so it would be great to at least ensure compatibility with the latest release.

**Comment by EnricoMi** [comment]
Will do.

**Comment by maxhgerlach** [comment]
I think `torch==2.0.0+cu118` should work, and I believe that package brings its own CUDA runtime libraries, so it may just work in the same container.

The TF 2.12 errors will hopefully be fixed via https://github.com/horovod/horovod/pull/3894.

**Comment by maxhgerlach** [comment]
Almost there, just need to update cuDNN to 8.6 for TensorFlow 2.12 ...
```
[0]<stderr>:2023-04-22 14:55:55.839805: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:417] Loaded runtime CuDNN library: 8.4.1 but source was compiled with: 8.6.0.  CuDNN library needs to have matching major version and equal or higher minor version. If using a binary install, upgrade your CuDNN library.  If building from sources, make sure the library loaded at runtime is compatible with the version specified during compile configuration.
```

**Comment by EnricoMi** [comment]
@maxhgerlach Upgraded CUDNN to 8.6.0, which is only available for CUDA 11.8, so upgrading base image and NCCL to CUDA 11.8 as well.

**Comment by EnricoMi** [comment]
Remaining TF 2.12 issue:

```
[1]<stdout>:E       3 root error(s) found.
[1]<stdout>:E         (0) UNKNOWN: ncclAllGather failed: unhandled cuda error
[1]<stdout>:E       	 [[{{node HorovodAllgather_Cast_0}}]]
[1]<stdout>:E       	 [[HorovodAllgather_Cast_0/_3]]
[1]<stdout>:E       	 [[Slice_1/_9]]
[1]<stdout>:E         (1) UNKNOWN: ncclAllGather failed: unhandled cuda error
[1]<stdout>:E       	 [[{{node HorovodAllgather_Cast_0}}]]
[1]<stdout>:E       	 [[HorovodAllgather_Cast_0/_3]]
[1]<stdout>:E         (2) UNKNOWN: ncclAllGather failed: unhandled cuda error
[1]<stdout>:E       	 [[{{node HorovodAllgather_Cast_0}}]]
[1]<stdout>:E       0 successful operations.
[1]<stdout>:E       0 derived errors ignored.
```

**Comment by maxhgerlach** [comment]
Thanks!

>Remaining TF 2.12 issue:

Probably related to this CUDA/NCCL problem again, from the Buildkite logs:

```
misc/strongstream.cc:32 NCCL WARN Cuda failure 'CUDA driver is a stub library'
```

This happens with the following settings, for both TF 2.12 and TF Head:
```
        CUDA_DOCKER_VERSION: 11.8.0-devel-ubuntu20.04
        CUDNN_VERSION: 8.6.0.163-1+cuda11.8
        NCCL_VERSION_OVERRIDE: 2.15.1-1+cuda11.8
```

On master we're still using the older NCCL package (https://github.com/horovod/horovod/blob/master/docker-compose.test.yml#L224) with tf-head, so that seems to work for now:
```
        CUDA_DOCKER_VERSION: 11.8.0-devel-ubuntu20.04
        CUDNN_VERSION: 8.6.0.163-1+cuda11.8
        NCCL_VERSION_OVERRIDE: 2.11.4-1+cuda11.6
```

But this is not great. I thought the 'CUDA driver is a stub library' error would have gone away with https://github.com/horovod/horovod/pull/3867.

**Comment by maxhgerlach** [comment]
I've attached the TF 2.12 GPU container build log from Buildkite here: [build_log_test-gpu-openmpi-gloo-py3_8-tf2_12_0-keras2_12_0-torch2_0_0-mxnet1_9_1-pyspark3_4_0.txt](https://github.com/horovod/horovod/files/11310424/build_log_test-gpu-openmpi-gloo-py3_8-tf2_12_0-keras2_12_0-torch2_0_0-mxnet1_9_1-pyspark3_4_0.txt)

From that log, this is the linker command line for `horovod/tensorflow/mpi_lib`:
```
[0m[91m  [100%] Linking CXX shared library /tmp/pip-req-build-0oq52ric/build/lib.linux-x86_64-3.8/horovod/tensorflow/mpi_lib.cpython-38-x86_64-linux-gnu.so

  cd /tmp/pip-req-build-0oq52ric/build/temp.linux-x86_64-3.8/RelWithDebInfo/horovod/tensorflow && /usr/local/lib/python3.8/dist-packages/cmake/data/bin/cmake -E cmake_link_script CMakeFiles/tensorflow.dir/link.txt --verbose=1

[0m[91m  /usr/bin/c++ -fPIC -I/usr/local/lib/python3.8/dist-packages/tensorflow/include -D_GLIBCXX_USE_CXX11_ABI=1 --std=c++17 -DEIGEN_MAX_ALIGN_BYTES=64  -pthread -fPIC -Wall -ftree-vectorize -mf16c -mavx -mfma -O3 -g -DNDEBUG  -Wl,--version-script=/tmp/pip-req-build-0oq52ric/horovod.lds -Wl,-Bsymbolic-functions -Wl,-z,relro,-z,now -shared -Wl,-soname,mpi_lib.cpython-38-x86_64-linux-gnu.so -o /tmp/pip-req-build-0oq52ric/build/lib.linux-x86_64-3.8/horovod/tensorflow/mpi_lib.cpython-38-x86_64-linux-gnu.so CMakeFiles/tensorflow.dir/__/common/common.cc.o CMakeFiles/tensorflow.dir/__/common/controller.cc.o CMakeFiles/tensorflow.dir/__/common/fusion_buffer_manager.cc.o CMakeFiles/tensorflow.dir/__/common/group_table.cc.o CMakeFiles/tensorflow.dir/__/common/half.cc.o CMakeFiles/tensorflow.dir/__/common/logging.cc.o CMakeFiles/tensorflow.dir/__/common/message.cc.o CMakeFiles/tensorflow.dir/__/common/operations.cc.o CMakeFiles/tensorflow.dir/__/common/parameter_manager.cc.o CMakeFiles/tensorflow.dir/__/common/process_set.cc.o CMakeFiles/tensorflow.dir/__/common/response_cache.cc.o CMakeFiles/tensorflow.dir/__/common/stall_inspector.cc.o CMakeFiles/tensorflow.dir/__/common/thread_pool.cc.o CMakeFiles/tensorflow.dir/__/common/timeline.cc.o CMakeFiles/tensorflow.dir/__/common/tensor_queue.cc.o CMakeFiles/tensorflow.dir/__/common/ops/collective_operations.cc.o CMakeFiles/tensorflow.dir/__/common/ops/operation_manager.cc.o CMakeFiles/tensorflow.dir/__/common/optim/bayesian_optimization.cc.o CMakeFiles/tensorflow.dir/__/common/optim/gaussian_process.cc.o CMakeFiles/tensorflow.dir/__/common/utils/env_parser.cc.o CMakeFiles/tensorflow.dir/__/common/mpi/mpi_context.cc.o CMakeFiles/tensorflow.dir/__/common/mpi/mpi_controller.cc.o CMakeFiles/tensorflow.dir/__/common/ops/mpi_operations.cc.o CMakeFiles/tensorflow.dir/__/common/ops/adasum/adasum_mpi.cc.o CMakeFiles/tensorflow.dir/__/common/ops/adasum_mpi_operations.cc.o CMakeFiles/tensorflow.dir/__/common/ops/cuda_operations.cc.o CMakeFiles/tensorflow.dir/__/common/ops/gpu_operations.cc.o CMakeFiles/tensorflow.dir/__/common/ops/mpi_gpu_operations.cc.o CMakeFiles/tensorflow.dir/__/common/ops/nccl_operations.cc.o CMakeFiles/tensorflow.dir/__/common/nvtx_op_range.cc.o CMakeFiles/tensorflow.dir/__/common/gloo/gloo_context.cc.o CMakeFiles/tensorflow.dir/__/common/gloo/gloo_controller.cc.o CMakeFiles/tensorflow.dir/__/common/gloo/http_store.cc.o CMakeFiles/tensorflow.dir/__/common/gloo/memory_store.cc.o CMakeFiles/tensorflow.dir/__/common/ops/gloo_operations.cc.o CMakeFiles/tensorflow.dir/__/common/ops/adasum_gpu_operations.cc.o CMakeFiles/tensorflow.dir/mpi_ops.cc.o CMakeFiles/tensorflow.dir/xla_mpi_ops.cc.o   -L/usr/local/cuda/targets/x86_64-linux/lib  -Wl,-rpath,/usr/local/lib /usr/local/lib/libmpi.so /usr/lib/x86_64-linux-gnu/libnccl_static.a -ldl -L/usr/local/lib/python3.8/dist-packages/tensorflow -l:libtensorflow_framework.so.2 /usr/local/lib/python3.8/dist-packages/tensorflow/python/_pywrap_tensorflow_internal.so /usr/local/lib/python3.8/dist-packages/tensorflow/libtensorflow_cc.so.2 ../../third_party/compatible17_gloo/gloo/libcompatible17_gloo.a ../common/ops/cuda/libhorovod_cuda_kernels.a /usr/local/lib/libmpi.so -lpthread -lcudadevrt -lcudart_static -lrt -lpthread -ldl
```

As far as I can tell we're linking both NCCL and CUDA RT statically here (`-lcudart_static`, `/usr/lib/x86_64-linux-gnu/libnccl_static.a`). NCCL comes from apt packages `libnccl-dev 2.15.1-1+cuda11.8` and `libnccl2 2.15.1-1+cuda11.8`.

**Comment by maxhgerlach** [comment]
Do we require this CUDA stubs related line in `Dockerfile.test.gpu` to build Horovod? https://github.com/horovod/horovod/blob/7be9a001528dafbf7351bbb48062d1319d856967/Dockerfile.test.gpu#L212

**Comment by maxhgerlach** [comment]
Updating NCCL to 2.16.5 didn't make a difference in the pipeline.

> Is the base image misconfigured? Or will that LIBRARY_PATH be modified throughout the build process?

I'm also not sure about the effect of these.

Something weird is going on for sure. Please observe the following local interaction:

```
horovod$ git log -n 1
commit 23908b94ea88381b2918647b883de5d71106e717 (HEAD -> branch-ci-upgrade-test-frameworks-8, upstream/branch-ci-upgrade-test-frameworks-8)
Author: Max H. Gerlach <git@maxgerlach.de>
Date:   Mon Apr 24 17:06:45 2023 +0200

    Try NCCL 2.16.5-1+cuda11.8

    Signed-off-by: Max H. Gerlach <git@maxgerlach.de> 

horovod$ docker-compose -f docker-compose.test.yml build --pull test-gpu-openmpi-gloo-py3_8-tf2_12_0-keras2_12_0-torch2_0_0-mxnet1_9_1-pyspark3_4_0 2>&1 | tee build_locally_test-gpu-openmpi-gloo-py3_8-tf2_12_0-keras2_12_0-torch2_0_0-mxnet1_9_1-pyspark3_4_0.txt
# Docker build is successful

horovod$ docker run -it --gpus all horovod_test-gpu-openmpi-gloo-py3_8-tf2_12_0-keras2_12_0-torch2_0_0-mxnet1_9_1-pyspark3_4_0:latest
root@0bf5cfedcb7e:/# cd /horovod/test/parallel/ && OMPI_ALLOW_RUN_AS_ROOT=1 OMPI_ALLOW_RUN_AS_ROOT_CONFIRM=1 mpirun -np 2 pytest -svx test_tensorflow.py -k test_horovod_allreduce_average_gpu 2>&1 | grep -e 'CUDA driver is a stub library' -e '===='
============================= test session starts ==============================
============================= test session starts ==============================
0bf5cfedcb7e:138:225 [0] misc/strongstream.cc:60 NCCL WARN Cuda failure 'CUDA driver is a stub library'
=================================== FAILURES ===================================
=============================== warnings summary ===============================
=========================== short test summary info ============================
=================================== FAILURES ===================================
================ 1 failed, 144 deselected, 2 warnings in 10.57s ================
=============================== warnings summary ===============================
=========================== short test summary info ============================
================ 1 failed, 144 deselected, 2 warnings in 10.58s ================    
# test fails like in the CI pipeline

# rebuild Horovod with command line copied from Dockerfile
root@0bf5cfedcb7e:/horovod# cd /horovod && \
>     python setup.py sdist && \
>     ldconfig /usr/local/cuda/targets/x86_64-linux/lib/stubs && \
>     bash -c "${HOROVOD_BUILD_FLAGS} HOROVOD_WITH_TENSORFLOW=1 HOROVOD_WITH_PYTORCH=1 HOROVOD_WITH_MXNET=1 pip install --no-cache-dir -v $(ls /horovod/dist/horovod-*.tar.gz)[spark,ray]" && \
>     ldconfig
# ... build succeeds

root@0bf5cfedcb7e:/horovod# cd /horovod/test/parallel/ && OMPI_ALLOW_RUN_AS_ROOT=1 OMPI_ALLOW_RUN_AS_ROOT_CONFIRM=1 mpirun -np 2 pytest -svx test_tensorflow.py -k test_horovod_allreduce_average_gpu 2>&1 | grep -e 'CUDA driver is a stub library' -e '===='
============================= test session starts ==============================
============================= test session starts ==============================
=============================== warnings summary ===============================
================ 1 passed, 144 deselected, 2 warnings in 12.74s ================
=============================== warnings summary ===============================
================ 1 passed, 144 deselected, 2 warnings in 12.74s ================
# test succeeds this time!
```
- I build the test image locally and run the container.
- Inside the container I run a test, which fails with 'CUDA driver is a stub library'.
- Then, still inside the same container I build Horovod again without any further changes.
- I rerun the same test, this time it succeeds.


=> So is there some kind of mismatch between the environments at container build and run time?

**Comment by maxhgerlach** [comment]
To summarize:

- When we build Horovod in the CI with CUDA 11.8, but NCCL **2.11.4-1+cuda11.6**, tests run fine.
- When we build Horovod in the CI with CUDA 11.8 and NCCL **2.16.5-1+cuda11.8**, we get `'CUDA driver is a stub library'` errors at (GPU) container run time.
    - But with the same libraries, when we build Horovod in the running GPU container, tests run fine again.

@cliffwoolley, do you maybe have an idea what could be going on here? It's different from the situation in #3831: Horovod links both NCCL and CUDA RT statically now.

**Comment by EnricoMi** [comment]
> when we build Horovod in the running GPU container

What do you mean with `in the running GPU container`?

**Comment by maxhgerlach** [comment]
Sorry for the wonky terminology! I meant inside `docker run --gpus all ...` (rather than during the docker / buildkit build); I guess that's with CUDA libraries provided by the GPU driver, and not just stubs.

**Comment by maxhgerlach** [comment]
>You say the GPU machines provide some drivers that are missing during build?

More or less. I think that's what the 'CUDA driver is a stub library' may be about (here's a brief explanation regarding these stubs: https://stackoverflow.com/a/52784819/1095888). It should be possible to link Horovod / NCCL correctly with stubs, then run the resulting container with the actual CUDA driver.

(This is not Buildkite specific. I observed the same problem with a local GPU server.)

**Comment by EnricoMi** [comment]
@romerojosh does https://github.com/horovod/horovod/pull/3846#issuecomment-1521392056 look familiar?

**Comment by maxhgerlach** [comment]
My last commit fixes the issue in my local test environment, although it's a bit of a hack. Let's see if the CI passes now.

More detailed explanation:

The 'CUDA driver is a stub library' error message was a bit misleading. The problem wasn't actually related to an accidental usage of CUDA driver stubs, but again a similar manifestation of the underlying cause of issue https://github.com/horovod/horovod/issues/3831.

Starting with about version 2.12 NCCL comes with an "enhanced compatibility" shim [enhcompat.cc](https://github.com/NVIDIA/nccl/blob/master/src/enhcompat.cc). I think it's meant to improve compatibility in case NCCL had been built with a newer version of the CUDA runtime than what is available when linking NCCL to an application, but I'm not entirely sure how that works. Anyway, it reuses `cudaErrorStubLibrary = 34`.

`enhcompat.cc` contains a couple of weak symbols that are meant to be overridden at link time by strong symbols in `libcudart_static.a`. However, when linking static libraries that only works by default if `-lcudart_static` appears on the linker command line _before_ `libnccl_static.a`, else the difference between weak/strong symbols is ignored (surprisingly, I think). This Stackoverflow answer explains that fact quite well: https://stackoverflow.com/a/37191811/1095888

Unfortunately, CMake puts `-lcudart_static` near the end of the command line automatically (via the builtin CUDA language support), and that can't be controlled easily. So my workaround is to put it in a second time directly after `libnccl_static.a`, this time wrapped by flags `Wl,--whole-archive` and `-Wl,--no-whole-archive`. Then its strong symbols properly replace the weak symbols encountered earlier by the linker.

File: CMakeLists.txt [hunk 1]
```diff
@@ -233,6 +233,10 @@ if(HOROVOD_GPU_ALLREDUCE STREQUAL "N" OR HOROVOD_GPU_ALLGATHER STREQUAL "N" OR H
         endif()
         include_directories(SYSTEM ${NCCL_INCLUDE_DIRS})
         list(APPEND LINKER_LIBS ${NCCL_LIBRARIES})
+        if(NCCL_LIBRARY_FILE_NAME MATCHES "static" AND lowercase_CMAKE_CUDA_RUNTIME_LIBRARY STREQUAL "static")
+            # ensure that weak symbols from NCCL's enhcompat.cc are properly overwritten by symbols from libcudart_static.a
+            list(APPEND LINKER_LIBS -Wl,--whole-archive cudart_static -Wl,--no-whole-archive)
```
Comment (EnricoMi) [inline CMakeLists.txt:238]: Can you add a link to this PR please when this hack works?

Comment (maxhgerlach) [reply CMakeLists.txt:238]: good idea, will do once the CI really passes

**Comment by maxhgerlach** [comment]
For completeness:

1)
> The build log says:
> 
> ```
> LIBRARY_PATH=/usr/local/cuda/lib64/stubs
> ```
> 
> Is the base image misconfigured? Or will that `LIBRARY_PATH` be modified throughout the build process?

Unsetting this in the Dockerfile when building Horovod didn't change anything.

2)
> Do we actually require this CUDA stubs related line in `Dockerfile.test.gpu` to build Horovod?
> 
> https://github.com/horovod/horovod/blob/7be9a001528dafbf7351bbb48062d1319d856967/Dockerfile.test.gpu#L212

Without this line building Horovod for MXnet fails. During `FindMxnet.cmake` a (stub or driver) libcuda.so is required to be present. (But removing the line also didn't fix the issue anyway.)

**Comment by EnricoMi** [comment]
Non-head GPU is green! :partying_face: I'd say this is ready for review.

**Comment by maxhgerlach** [comment]
@nvcastet, this comes with a change how NCCL and `libcudart_static` are linked. Does this look reasonable to you?

**Review by maxhgerlach — APPROVED** [review]

**Comment by nvcastet** [comment]
@maxhgerlach Thanks for taking the time to investigate the issue and finding a WAR :
> when linking static libraries that does not work by default if -lcudart_static appears on the linker command line before libnccl_static.a, else the difference between weak/strong symbols is ignored

> Unfortunately, CMake puts -lcudart_static near the end of the command line automatically 

By default, does CMake issue a command line with `-lcudart_static` before or after `-lnccl_static`?

**Comment by maxhgerlach** [comment]
> By default, does CMake issue a command line with -lcudart_static before or after -lnccl_static?

`-lnccl_static` is one of the libraries we put into the command line explicitly. In contrast `-lcudart_static` is put in  automatically (this is triggered by `enable_language(CUDA)`). This default appearance `-lcudart_static` is behind the user-added libraries, so in particular after `-lnccl_static`.

> when linking static libraries that does not work by default if -lcudart_static appears on the linker command line before libnccl_static.a, else the difference between weak/strong symbols is ignored

Actually, while I wrote that yesterday, I'm not sure if the reversed order `-lcudart_static -lnccl_static` would really work without `-Wl,--whole-archive`. Maybe the linker wouldn't see a reason to put in the strong symbols from cudart_static then before it encounters the weak symbols from nccl_static..

**Comment by maxhgerlach** [comment]
Thanks for giving this an extra pair of eyes, @nvcastet!

**Comment by nvcastet** [comment]
Thanks for the clarification @maxhgerlach. The use of weak/strong symbols seems to be really fragile.
We will internally discuss to figure out if we can improve NCCL to have a better solution.

**Comment by EnricoMi** [comment]
@maxhgerlach Looks like we have lost all MXnet2 tests during your last commit. Given there were no functional changes, I think this is transient, but I am investigating this first.

**Comment by EnricoMi** [comment]
Rerunning CPU head build and test as build timed out, the only place where we run MXnet 2 tests.

**Comment by EnricoMi** [comment]
Looks like tf-nightly just broke our head build:

```
2023-04-27T19:11:52.0603588Z #41 345.7   /tmp/pip-req-build-hywxpjt9/horovod/tensorflow/mpi_ops.cc: In function ‘horovod::common::Status horovod::tensorflow::{anonymous}::ConvertStatus(const Status&)’:
2023-04-27T19:11:52.0604582Z #41 345.7   /tmp/pip-req-build-hywxpjt9/horovod/tensorflow/mpi_ops.cc:134:48: error: ‘const Status’ {aka ‘const class absl::lts_20230125::Status’} has no member named ‘error_message’
2023-04-27T19:11:52.2116853Z #41 345.7     134 |     return common::Status::UnknownError(status.error_message());
2023-04-27T19:11:52.2117293Z #41 345.7         |                                                ^~~~~~~~~~~~~
2023-04-27T19:11:52.2118368Z #41 345.7   /tmp/pip-req-build-hywxpjt9/horovod/tensorflow/mpi_ops.cc:136:53: error: ‘const Status’ {aka ‘const class absl::lts_20230125::Status’} has no member named ‘error_message’
2023-04-27T19:11:52.2118909Z #41 345.7     136 |     return common::Status::PreconditionError(status.error_message());
2023-04-27T19:11:52.2119249Z #41 345.7         |                                                     ^~~~~~~~~~~~~
2023-04-27T19:11:52.2120108Z #41 345.7   /tmp/pip-req-build-hywxpjt9/horovod/tensorflow/mpi_ops.cc:138:43: error: ‘const Status’ {aka ‘const class absl::lts_20230125::Status’} has no member named ‘error_message’
2023-04-27T19:11:52.2120741Z #41 345.7     138 |     return common::Status::Aborted(status.error_message());
2023-04-27T19:11:52.2121067Z #41 345.7         |                                           ^~~~~~~~~~~~~
2023-04-27T19:11:52.2121847Z #41 345.7   /tmp/pip-req-build-hywxpjt9/horovod/tensorflow/mpi_ops.cc:140:51: error: ‘const Status’ {aka ‘const class absl::lts_20230125::Status’} has no member named ‘error_message’
2023-04-27T19:11:52.2122402Z #41 345.7     140 |     return common::Status::InvalidArgument(status.error_message());
2023-04-27T19:11:52.2122760Z #41 345.7         |                                                   ^~~~~~~~~~~~~
```

Which removes our MXNet 2 tests. Fixing HEAD is not objective of this PR.

**Comment by maxhgerlach** [comment]
Thanks for discussing these complications inside Nvidia, @nvcastet, and thanks for seeing this through @EnricoMi!

**Pull request merged**